### PR TITLE
Land Datagram Preferred (PR #342)

### PR DESCRIPTION
**Pull request by ianswett**
Update main to reflect the tx-mode branch after #342 was merged.

### CODE REVIEW

File: draft-ietf-moq-transport.md [hunk 1]
```diff
@@ -931,6 +935,36 @@ the receiver MUST close the session with a Protocol Violation.
...
+be the first message on a unidirectional stream.
+
+An Object received in an `OBJECT_PREFER_DATAGRAM` message has an `Object
+Forwarding Preference` = `Datagram`.
+
+To send an Object with `Object Forwarding Preference` = `Datagram`, determine
+the length of the fields and payload, and compare the length with the maximum
+datagram size of the session.  If the object size is less than or equal maximum
+datagram size, send the serialized data as a datagram.  Otherwise, open a
+stream, send the serialized data and terminate the stream.  An implementation
```
Comment (suhasHere) [inline draft-ietf-moq-transport.md:952]: i think we need to preserve the fact that this is preferred as datagram when we put it in a stream . Wonder where should that go ? 

Should we extend OBJECT_STREAM message with a bool saying "preferred_datagram". This is needed for receviers handling and forwarding

**Review by suhasHere — COMMENTED** [review]
Looks fine . Have one comment though

File: draft-ietf-moq-transport.md (hunk 1 above)
Comment (suhasHere) [reply draft-ietf-moq-transport.md:952]: this can be added in a separate PR too

**Review by suhasHere — APPROVED** [review]

File: draft-ietf-moq-transport.md (hunk 1 above)
Comment (afrind) [reply draft-ietf-moq-transport.md:952]: Preserving the preference was the intent.  When datagram is sent on a stream, it starts with OBJECT_PREFER_DATAGRAM (0x1) to communicate this.

File: draft-ietf-moq-transport.md [hunk 2]
```diff
@@ -931,6 +935,36 @@ the receiver MUST close the session with a Protocol Violation.
 
 * Other fields: As described in {{canonical-object-fields}}.
 
+**Object Prefer Datagram Message**
+
+An `OBJECT_PREFER_DATAGRAM` message carries a single object in a datagram or
+a stream. There is no explicit length of the payload; it is determined by the
+length of the datagram or stream.  If this message appears on a stream, it MUST
+be the first message on a unidirectional stream.
```
Comment (afrind) [inline draft-ietf-moq-transport.md:943]: s/first/only/  based on a comment in 342.

**Review by afrind — COMMENTED** [review]
Individual Review: 

There was some feedback on previous iterations about this PR lacking guidance when encountering flow control limits (eg: send as datagram, relay converts to stream, but is out of streams or flow control)

File: draft-ietf-moq-transport.md (hunk 2 above)
Comment (ianswett) [reply draft-ietf-moq-transport.md:943]:
```suggestion
be the only message on a unidirectional stream.
```

**Review by martinduke — COMMENTED** [review]
Whoa, I've missed some PRs, but LGTM

**Comment by ianswett** [comment]
> Individual Review:
> 
> There was some feedback on previous iterations about this PR lacking guidance when encountering flow control limits (eg: send as datagram, relay converts to stream, but is out of streams or flow control)

Agreed, but I'm not sure there's any normative language there.  Is this a case of writing a paragraph cautioning that such limits could cause issues?

**Pull request merged**